libraries/tickv/src/async_ops.rs

```diff
@@ -0,0 +1,923 @@
+//!         if self.async_erase_region.get() != region_number {
+//!             // We aren't ready yet, launch the async operation
+//!             self.async_erase_region.set(region_number);
+//!             return Err(ErrorCode::EraseNotReady(region_number));
+//!         }
+//!
+//!         Ok(())
+//!     }
+//! }
+//!
+//! // Create the TicKV instance and loop until everything is done
+//! // NOTE in an real implementation you will want to wait on
+//! // callbacks/interrupts and make this async.
+//!
+//! let mut read_buf: [u8; 1024] = [0; 1024];
+//! let tickv = AsyncTicKV::<FlashCtrl, DefaultHasher, 1024>::new(FlashCtrl::new(),
+//!                   &mut read_buf, 0x1000);
+//!
+//! let mut ret = tickv.initalise((&mut DefaultHasher::new(), &mut DefaultHasher::new()));
+//! while ret.is_err() {
+//!     // There is no actual delay here, in a real implementation wait on some event
+//!     ret = tickv.continue_operation(
+//!         (&mut DefaultHasher::new(), &mut DefaultHasher::new())).0;
+//!
+//!     match ret {
+//!         Err(ErrorCode::ReadNotReady(reg)) => {
+//!             tickv.set_read_buffer(&tickv.tickv.controller.buf.borrow()[reg]);
+//!         }
+//!         Ok(_) => break,
+//!         Err(ErrorCode::WriteNotReady(reg)) => break,
+//!         Err(ErrorCode::EraseNotReady(reg)) => {}
+//!         _ => unreachable!(),
+//!     }
+//! }
+//!
+//! // Then when calling the TicKV function check for the error. For example
+//! // when appending a key:
+//!
+//! // Add a key
+//! static VALUE: [u8; 32] = [0x23; 32];
+//! let ret = unsafe { tickv.append_key(&mut DefaultHasher::new(), b"ONE", &VALUE) };
+//!
+//! match ret {
+//!     Err(ErrorCode::ReadNotReady(reg)) => {
+//!         // There is no actual delay in the test, just continue now
+//!         tickv.set_read_buffer(&tickv.tickv.controller.buf.borrow()[reg]);
+//!         tickv
+//!             .continue_operation((&mut DefaultHasher::new(), &mut DefaultHasher::new())).0
+//!             .unwrap();
+//!     }
+//!     Ok(_) => {}
+//!     _ => unreachable!(),
+//! }
+//!
+//! ```
+//!
+//! This will call into the `FlashController` again where the
+//! `FlashController` implementation must return the data that is requested.
+//! If the data isn't ready (multiple reads might occur) then the `NotReady`
+//! error types can still be used.
+//!
+
+use crate::error_codes::ErrorCode;
+use crate::flash_controller::FlashController;
+use crate::success_codes::SuccessCode;
+use crate::tickv::{State, TicKV};
+use core::cell::Cell;
+use core::hash::Hasher;
+
+/// The return type from the continue operation
+type ContinueReturn = (
+    // Result
+    Result<SuccessCode, ErrorCode>,
+    // Buf Buffer
+    Option<&'static mut [u8]>,
+);
+
+/// The struct storing all of the TicKV information for the async implementation.
+pub struct AsyncTicKV<'a, C: FlashController<S>, H: Hasher, const S: usize> {
+    /// The controller used for flash commands
```

Comment:
I agree that it can be removed with the HIL. Whatever order it all ends up being approved I can rebase from there.

There will be more changes required to match the new HIL, but that should be easy to address.